File: python/src/heaton_life/ca/elementary.py
"""Elementary (Wolfram) cellular automata — spec/elementary.md.

State is the 1-D tape; the frame is the scrolling space-time diagram, which is
the canonical way to *look* at these — a clean example of state != frame.
"""

from __future__ import annotations

import dataclasses

import numpy as np
from numpy.typing import NDArray

from heaton_life.core.params import Params
from heaton_life.init import soup


@dataclasses.dataclass(frozen=True)
class ElementaryParams(Params):
    rule: int = dataclasses.field(default=30, metadata={"min": 0, "max": 255})
    width: int = dataclasses.field(default=256, metadata={"min": 8, "max": 4096})
    height: int = dataclasses.field(
        default=256, metadata={"min": 8, "max": 4096, "label": "Diagram Rows"}
    )
    init: str = dataclasses.field(default="single", metadata={"choices": ["single", "soup"]})
    density: float = dataclasses.field(
        default=0.5, metadata={"min": 0.0, "max": 1.0, "step": 0.01}
    )
    seed: int = dataclasses.field(
        default=0, metadata={"min": 0, "max": 4294967295, "role": "seed"}
    )
    boundary: str = dataclasses.field(default="torus", metadata={"choices": ["torus", "dead"]})


class Elementary:
    """An elementary CA. State is the uint8 0/1 tape, shape (width,)."""

    def __init__(
        self,
        rule: int = 30,
        *,
        size: tuple[int, int] = (256, 256),
        init: str | NDArray[np.uint8] = "single",
        density: float = 0.5,
        seed: int = 0,
        boundary: str = "torus",
    ) -> None:
        if not 0 <= rule <= 255:
            raise ValueError(f"rule must be 0..255, got {rule}")
        width, height = size
        self._initial: NDArray[np.uint8] | None = None
        if isinstance(init, np.ndarray):
            if init.shape != (width,):
                raise ValueError(f"init tape shape {init.shape} does not match (width={width},)")
            self._initial = (init > 0).astype(np.uint8)
            init_name = "array"
        else:
            init_name = init
        self.params = ElementaryParams(
            rule=rule,
            width=width,
            height=height,
            init=init_name,
            density=density,
            seed=seed,
            boundary=boundary,
        )
        # Rule table indexed by (left<<2 | center<<1 | right).
        self._table = np.array([(rule >> i) & 1 for i in range(8)], dtype=np.uint8)
        self._generation = 0
        self._tape: NDArray[np.uint8]
        self._diagram: NDArray[np.uint8]
        self.reset()

    @classmethod
    def from_params(cls, params: ElementaryParams) -> Elementary:
        if params.init == "array":
            raise ValueError("params with init='array' need the tape: Elementary(init=...)")
        return cls(
            params.rule,
            size=(params.width, params.height),
            init=params.init,
            density=params.density,
            seed=params.seed,
            boundary=params.boundary,
        )

    def reset(self, seed: int | None = None) -> None:
        if seed is not None:
            self.params = self.params.replace(seed=seed)
        p = self.params
        if p.init == "array":
            assert self._initial is not None
            tape = self._initial.copy()
        elif p.init == "single":
            tape = np.zeros(p.width, dtype=np.uint8)
            tape[p.width // 2] = 1
        elif p.init == "soup":
            tape = soup((p.width, 1), density=p.density, seed=p.seed)[0]
        else:
            raise ValueError(f"unknown init strategy: {p.init!r}")
        self._tape = tape
        self._diagram = np.zeros((p.height, p.width), dtype=np.uint8)
        self._diagram[0] = tape
        self._generation = 0

    def step(self, n: int = 1) -> None:
        p = self.params
        for _ in range(n):
            tape = self._tape
            if p.boundary == "torus":
                left = np.roll(tape, 1)
                right = np.roll(tape, -1)
            else:  # dead
                left = np.zeros_like(tape)
                left[1:] = tape[:-1]
                right = np.zeros_like(tape)
                right[:-1] = tape[1:]
            index = (left << 2) | (tape << 1) | right
            self._tape = self._table[index]
            self._generation += 1
            row = self._generation
            if row < p.height:
                self._diagram[row] = self._tape
            else:  # scroll up
                self._diagram[:-1] = self._diagram[1:]
                self._diagram[-1] = self._tape

    @property
    def state(self) -> NDArray[np.uint8]:
        return self._tape

    @property
    def generation(self) -> int:
        return self._generation

    def frame(self) -> NDArray[np.uint8]:
        return self._diagram * np.uint8(255)

File: dotnet/src/HeatonLife.Core/Elementary.cs
using System;

namespace HeatonLife
{
    /// <summary>
    /// Elementary (Wolfram) cellular automaton (spec/elementary.md). State is the 1-D
    /// 0/1 tape, Width bytes; the diagram is the scrolling space-time record (Height
    /// rows, row-major 0/1) — the canonical way to *look* at these. Bit-exact tier.
    /// </summary>
    public sealed class Elementary : IIndexedFrameSource
    {
        private readonly byte[] _table; // indexed by (left<<2 | center<<1 | right)
        private byte[] _tape;
        private byte[] _scratch;
        private readonly byte[] _diagram;

        public int Rule { get; }
        public int Width { get; }
        public int Height { get; }
        public Boundary Boundary { get; }
        public int Generation { get; private set; }

        /// <summary>The current tape (Width bytes, 0/1).</summary>
        public ReadOnlySpan<byte> State => _tape;

        /// <summary>The space-time diagram (Height*Width bytes, row-major 0/1).</summary>
        public ReadOnlySpan<byte> Diagram => _diagram;

        public Elementary(int rule, int width, int height, Boundary boundary = Boundary.Torus)
        {
            if (rule < 0 || rule > 255)
                throw new ArgumentOutOfRangeException(nameof(rule), $"rule must be 0..255, got {rule}");
            Rule = rule;
            Width = width;
            Height = height;
            Boundary = boundary;
            _table = new byte[8];
            for (int i = 0; i < 8; i++)
                _table[i] = (byte)((rule >> i) & 1);
            _tape = new byte[width];
            _scratch = new byte[width];
            _diagram = new byte[width * height];
            SeedSingle();
        }

        /// <summary>Single live cell at Width / 2. Consumes no RNG draws.</summary>
        public void SeedSingle()
        {
            Array.Clear(_tape, 0, _tape.Length);
            _tape[Width / 2] = 1;
            ResetDiagram();
            Remember(_seed, _ =>
            {
                Array.Clear(_tape, 0, _tape.Length);
                _tape[Width / 2] = 1;
                ResetDiagram();
            });
        }

        /// <summary>Soup init per spec: one PCG32 draw per cell, alive iff draw &lt; floor(density * 2^32).</summary>
        public void SeedSoup(double density, uint seed)
        {
            Seeding.Soup(_tape, density, seed);
            ResetDiagram();
            Remember(seed, s => { Seeding.Soup(_tape, density, s); ResetDiagram(); });
        }

        // --- reset ---------------------------------------------------------------
        // How this world was last seeded, so Reset can replay it. The Python
        // reference keeps init/density/seed on its params object and re-dispatches
        // in reset(); a captured delegate is the same idea without a params type.
        // Every replay ends in ResetDiagram(), so the space-time record restarts
        // from the fresh tape exactly as it does after a first seeding.
        private Action<uint> _replayInit = _ => { };
        private uint _seed;

        private void Remember(uint seed, Action<uint> replay)
        {
            _seed = seed;
            _replayInit = replay;
        }

        /// <inheritdoc />
        public void Reset(uint? seed = null)
        {
            if (seed.HasValue)
                _seed = seed.Value;
            _replayInit(_seed);
        }

        /// <summary>Load an explicit 0/1 tape (Width entries).</summary>
        public void SetState(ReadOnlySpan<byte> tape)
        {
            if (tape.Length != Width)
                throw new ArgumentException($"expected {Width} cells, got {tape.Length}");
            for (int i = 0; i < tape.Length; i++)
                _tape[i] = (byte)(tape[i] > 0 ? 1 : 0);
            ResetDiagram();
            var snapshot = _tape.Clone() as byte[];
            Remember(_seed, _ =>
            {
                Array.Copy(snapshot!, _tape, _tape.Length);
                ResetDiagram();
            });
        }

        /// <summary>
        /// Restore a saved tape at a given generation. The diagram restarts from the
        /// tape, placed in the row a world that had stepped to that generation would
        /// hold it in (so stepping on continues the picture downward rather than
        /// leaving a gap); the rows above it are gone. A host that saved the diagram
        /// should use the overload that takes it.
        /// </summary>
        public void SetState(ReadOnlySpan<byte> tape, int generation)
        {
            SetState(tape);
            int row = TapeRow(generation);
            if (row != 0)
            {
                Array.Clear(_diagram, 0, Width);
                Array.Copy(_tape, 0, _diagram, row * Width, Width);
            }
            Generation = generation;
        }

        /// <summary>
        /// Restore a saved tape together with its space-time diagram (Height*Width
        /// bytes, row-major 0/1) at a given generation. This is what a host that
        /// persists a world needs: the diagram is the record of the steps already
        /// taken and cannot be rebuilt from the tape (spec/elementary.md "State"), so
        /// a world restored from its tape alone reopens blank. Reset replays the tape
        /// from a fresh diagram, exactly as after the tape-only overload.
        /// </summary>
        public void SetState(ReadOnlySpan<byte> tape, ReadOnlySpan<byte> diagram, int generation)
        {
            if (diagram.Length != _diagram.Length)
                throw new ArgumentException(
                    $"expected {_diagram.Length} diagram cells, got {diagram.Length}");
            SetState(tape);
            for (int i = 0; i < diagram.Length; i++)
                _diagram[i] = (byte)(diagram[i] > 0 ? 1 : 0);
            Generation = generation;
        }

        /// <summary>The diagram row that holds the tape at a given generation.</summary>
        private int TapeRow(int generation) => Math.Min(Math.Max(generation, 0), Height - 1);

        private void ResetDiagram()
        {
            Array.Clear(_diagram, 0, _diagram.Length);
            Array.Copy(_tape, 0, _diagram, 0, Width);
            Generation = 0;
        }

        public void Step(int n = 1)
        {
            for (int s = 0; s < n; s++)
                StepOnce();
        }

        /// <summary>Frame per spec/render.md: the space-time diagram, palette index = cell * 255.</summary>
        public void WriteFrame(byte[] frame)
        {
            if (frame.Length != _diagram.Length)
                throw new ArgumentException($"expected {_diagram.Length} bytes, got {frame.Length}");
            for (int i = 0; i < _diagram.Length; i++)
                frame[i] = (byte)(_diagram[i] * 255);
        }

        /// <summary>Allocating convenience for <see cref="WriteFrame"/>.</summary>
        public byte[] Frame()
        {
            var frame = new byte[_diagram.Length];
            WriteFrame(frame);
            return frame;
        }

        private void StepOnce()
        {
            int w = Width;
            bool torus = Boundary == Boundary.Torus;
            for (int x = 0; x < w; x++)
            {
                int left = torus ? _tape[(x - 1 + w) % w] : (x > 0 ? _tape[x - 1] : 0);
                int right = torus ? _tape[(x + 1) % w] : (x < w - 1 ? _tape[x + 1] : 0);
                _scratch[x] = _table[(left << 2) | (_tape[x] << 1) | right];
            }
            (_tape, _scratch) = (_scratch, _tape);
            Generation++;
            int row = Generation;
            if (row < Height)
            {
                Array.Copy(_tape, 0, _diagram, row * w, w);
            }
            else // scroll up
            {
                Array.Copy(_diagram, w, _diagram, 0, (Height - 1) * w);
                Array.Copy(_tape, 0, _diagram, (Height - 1) * w, w);
            }
        }
    }
}
